src/data_structures/single_linked_list.rs

```diff
@@ -0,0 +1,248 @@
+    // The push method which the signature's first parameter is self
+    pub fn push(&mut self, elem: T) {
+        let new_node = Box::new(Node {
+            elem,
+            // mem::replace function lets us steal a value out of a borrow by replacing it with another value
+            next: self.head.take(),
+        });
+
+        self.head = Some(new_node);
+    }
+    ///
+    /// In pop function, we trying to:
+    /// * check if the list is empty, so we use enum Option<T>, it can either be Some(T) or None
+    ///   * if it's empty, return None
+    ///   * if it's not empty
+    ///     * remove the head of the list
+    ///     * remove its elem
+    ///     * replace the list's head with its next
+    ///     * return Some(elem), as the situation if need
+    ///
+    /// so, we need to remove the head, and return the value of the head
+    pub fn pop(&mut self) -> Option<T> {
+        match self.head.take() {
+            None => None,
+            Some(node) => {
+                self.head = node.next;
+                Some(node.elem)
+            }
+        }
+    }
+
+    pub fn peek(&self) -> Option<&T> {
+        // Converts from &Option<T> to Option<&T>.
+        match self.head.as_ref() {
+            None => None,
+            Some(node) => Some(&node.elem),
+        }
+    }
+
+    pub fn peek_mut(&mut self) -> Option<&mut T> {
+        match self.head.as_mut() {
+            None => None,
+            Some(node) => Some(&mut node.elem),
+        }
+    }
+}
+
+/// The drop method of single-linked-list. There's a question that do we need to worry about cleaning up our list?
+/// As we all know the ownership and borrow mechanism, so we know the type will clean automatically after it goes out the scope,
+/// this implement by the Rust compiler automatically did which mean add trait `drop` for the automatically.
+///
+/// So, the complier will implements Drop for `List->Link->Box<Node> ->Node` automatically and tail recursive to clean the elements
+/// one by one. And we know the recursive will stop at Box<Node>
+/// https://rust-unofficial.github.io/too-many-lists/first-drop.html
+///
+/// As we know we can't drop the contents of the Box after deallocating, so we need to manually write the iterative drop
+
+impl<T> Drop for SingleLinkedList<T> {
+    fn drop(&mut self) {
+        let mut cur_link = self.head.take();
+        while let Some(mut boxed_node) = cur_link {
+            cur_link = boxed_node.next.take();
+            // boxed_node goes out of scope and gets dropped here;
+            // but its Node's `next` field has been set to None
+            // so no unbound recursion occurs.
+        }
+    }
+}
+
+/// Rust has nothing like a yield statement, and there's actually 3 different kinds of iterator should to implement
+
+// Collections are iterated in Rust using the Iterator trait, we define a struct implement Iterator
+pub struct IntoIter<T>(SingleLinkedList<T>);
+
+impl<T> SingleLinkedList<T> {
```

Comment:
(here and below) is it really a good idea to use separate `impls X` in the same module?

Reply:
Here, I will group them into one basic `impl` above.